crates/matrix-sdk-crypto/src/verification/cache.rs

```diff
@@ -40,19 +40,47 @@ impl VerificationCache {
         self.verification.iter().all(|m| m.is_empty())
     }
 
+    /// Add a new `Verification` object to the cache, this will cancel any
+    /// duplicates we have going on with a given user.
     pub fn insert(&self, verification: impl Into<Verification>) {
         let verification = verification.into();
 
-        self.verification
-            .entry(verification.other_user().to_owned())
-            .or_default()
-            .insert(verification.flow_id().to_owned(), verification);
+        let entry = self.verification.entry(verification.other_user().to_owned()).or_default();
+        let user_verifications = entry.value();
+
+        // Cancel all the old verifications as well as the new one we have for
+        // this user if someone tries to have two verifications going on at
+        // once.
+        for old in user_verifications {
+            let old_verification = old.value();
+
+            if !old_verification.is_cancelled() {
+                if let Some(r) = old_verification.cancel() {
+                    self.add_request(r.into())
+                }
+
+                if let Some(r) = verification.cancel() {
+                    self.add_request(r.into())
+                }
+            }
+        }
+
+        user_verifications.insert(verification.flow_id().to_owned(), verification);
```

Comment:
I don't understand the reasoning behind storing the verification in case it was cancelled. What's the point of doing that?

Reply:
Well is there a point of throwing it away immediately? 

The user will see an event that should have triggered a new verification, if we throw the object away the user won't have a way of knowing that the verification was processed. If we don't throw the object away, they can know that we indeed processed the event but the verification got cancelled.

The object will be garbage collected once we receive the next sync.

Reply:
> If we don't throw the object away, they can know that we indeed processed the event but the verification got cancelled.

If you think that is needed, sure. We should document this on the API though, it is not intuitive that the request added is cancelled, too.

Reply:
Added an explanation in https://github.com/matrix-org/matrix-rust-sdk/pull/727/commits/bf91d1c79f14d7782105ece9b7caa65a67502c0c.